### PR TITLE
Add case control statement

### PR DESCRIPTION
**Pull request by cdepillabout**
This PR adds the `case` control statement.

It is a WIP because commit 3ee3585 currently only adds the parser.  The part that turns a `DocCase` expression into a `Q Exp` is not done yet.  I will add that next week.

### CODE REVIEW

**Comment by cdepillabout** [comment]
Commit a468710 fixes the build failures caused by older versions of GHC.  However, build failures are still occuring because of an incomplete pattern match (because the code for `DocCase` is not yet implemented).  This will be fixed next week.

**Comment by cdepillabout** [comment]
Commit 9912e4c adds code to `Text.Heterocephalus.docToExp` to handle the `case` control statement.  I basically just copied it from `shakespeare`'s `Text.Hamlet.docToExp` function.

File: src/Text/Heterocephalus.hs [hunk 1]
```diff
@@ -486,6 +486,17 @@ docToExp set scope (DocCond conds final) = do
...
       return $ TupE [d', docs']
+docToExp set scope (DocCase deref cases) = do
+    let exp_ = derefToExp scope deref
+    matches <- mapM toMatch cases
+    return $ CaseE exp_ matches
+  where
+    toMatch :: (Binding, [Doc]) -> Q Match
+    toMatch (idents, inside) = do
+        (pat, extraScope) <- bindingPattern idents
+        let scope' = extraScope ++ scope
```
Comment (cdepillabout) [inline src/Text/Heterocephalus.hs:497]: I don't yet understand %100 how `Scope` works with heterocephalus/shakespeare.

@arowM Does this look alright? Or does something need to be changed for heterocephalus?

I copied this branch of the `docToExp` verbatim from `Text.Hamlet`.

**Comment by cdepillabout** [comment]
Tomorrow I will add a commit fixing the documentation.

File: src/Text/Heterocephalus.hs (hunk 1 above)
Comment (arowM) [reply src/Text/Heterocephalus.hs:497]: @cdepillabout 
It looks good!

Comment (cdepillabout) [reply src/Text/Heterocephalus.hs:497]: Thanks!

**Comment by cdepillabout** [comment]
Commit 8aee80a adds information about the `case` control statement to the cabal file and README.

Commit 2a75555 modifies `Text.Heterocephalus.Parse.Control` to allow for operators to be used as data constructors.

This makes it possible to write something like this:

```
%{case [1,2,3] }
%{ of (:) h ts }
...
%{ endcase }
```

Without this commit, the parser would fail at `(:)`.

One problem with this is that the parsing code in `Text.Heterocephalus.Parse.Control` is both pretty complicated and not robust.  There are corner cases that it doesn't handle.

This PR is ready for review/merging.

**Comment by cdepillabout** [comment]
In the long term, I'm wondering if it would be a better idea to use [haskell-src-exts](https://hackage.haskell.org/package/haskell-src-exts) to parse the things that shakespeare and heterocephalus are calling `Deref`s and `Binding`s.

haskell-src-exts provides two functions we could use.  One is [`parseExp`](https://hackage.haskell.org/package/haskell-src-exts-1.19.0/docs/Language-Haskell-Exts-Parser.html#v:parseExp).  This parses valid Haskell expressions.  This corresponds to heterocephalus' `Deref`.

The other function is [`parsePat`](https://hackage.haskell.org/package/haskell-src-exts-1.19.0/docs/Language-Haskell-Exts-Parser.html#v:parsePat).  This parses valid Haskell patterns.  This corresponds to heteroceaphlus' `Binding`.

Switching heterocephalus to use haskell-src-exts would be a big change.  It would require some way to convert from haskell-src-exts data types to template-haskell data types.  This should be straight-forward, but it would take time to do.  However, it looks like there might be packages that already do this: [haskell-src-meta](https://hackage.haskell.org/package/haskell-src-meta).

It would also require some way to pull out `Scope`s from the haskell-src-exts data types.  I'm not sure how easy this would be.

**Comment by arowM** [comment]
@cdepillabout 
Thank you!

About `haskell-src-exts`, I feel that it does not have enough benefits to change lots of codes.
I guess the main benefit to use `haskell-src-exts` is to enable full Haskell expressions, but `Deref`s are not required to fully support Haskell expressions, because template syntax of `heterocephalus` is just a `View`. (Providing rich expressions can make `View` fat.)

If you think it is necessary to discuss more about adopting `haskell-src-exts`, please create an independent issue :)

**Comment by cdepillabout** [comment]
@arowM I agree with you for the most part.

The problem with using the shakespeare code is that it is pretty hacky.  It only supports a small subset of Haskell expressions and patterns.  It is difficult to both understand and modify.  If we could replace it with `haskell-src-exts`, then we could remove some of the `Command` parsing code from heterocephalus.  I think this would make heterocephalus both easier to maintain and more flexible to use.

However, I agree that there is not enough benefit to change heterocephalus to use `haskell-src-exts`.  It would require too much work for too little gain.

**Comment by arowM** [comment]
@cdepillabout 
It makes sense!

**Pull request merged**